File: Grind75/C++/numberOfIslands.cpp
class Solution {
public:
    // Directional offsets for navigating neighboring cells
    vector<int> dirs = {-1, 0, 1, 0, -1};

    // Helper function to perform BFS traversal from a given cell
    void bfs(int row, int col, vector<vector<char>>& grid, vector<vector<int>>& visited) {
        queue<pair<int, int>> q;
        q.push({row, col});
        visited[row][col] = 1;

        while (!q.empty()) {
            pair<int, int> current = q.front();
            q.pop();

            // Explore all four directions
            for (int i = 0; i < 4; i++) {
                int newRow = current.first + dirs[i];
                int newCol = current.second + dirs[i + 1];

                // Check if the new cell is within bounds and is unvisited land
                if (newRow >= 0 && newRow < grid.size() && newCol >= 0 && newCol < grid[0].size() &&
                    grid[newRow][newCol] == '1' && visited[newRow][newCol] == 0) {
                    // Mark the cell as visited and add it to the queue for further exploration
                    visited[newRow][newCol] = 1;
                    q.push({newRow, newCol});
                }
            }
        }
    }

    // Function to count the number of islands in the grid
    int numIslands(vector<vector<char>>& grid) {
        int rows = grid.size(), cols = grid[0].size();
        vector<vector<int>> visited(rows, vector<int>(cols, 0)); // Initialize visited array

        int islandCount = 0; // Counter for the number of islands

        // Iterate through each cell in the grid
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                // If the cell represents unvisited land, perform BFS traversal from that cell
                if (grid[i][j] == '1' && visited[i][j] == 0) {
                    bfs(i, j, grid, visited);
                    islandCount++; // Increment island count after exploring all connected land cells
                }
            }
        }

        return islandCount;
    }
};

File: numberOfIslands.py
from collections import deque
from typing import List

class Solution:
    def bfs(self, i, j, visited, grid):
        row,col = len(grid), len(grid[0])
        dirs = [-1,0,1,0,-1]
        qq = deque()
        qq.append((i,j))
        visited[i][j] = True

        while qq:
            pair = qq.popleft()
            r = pair[0]
            c = pair[1]
            for d in range(4):
                r1 = r + dirs[d]
                c1 = c + dirs[d+1]
                if r1 >= 0 and r1 < row and c1 >= 0 and c1 < col and grid[r1][c1] == '1' and visited[r1][c1] == False:
                    qq.append((r1,c1))
                    visited[r1][c1] = True


    def numIslands(self, grid: List[List[str]]) -> int:
        row,col = len(grid), len(grid[0])
        count = 0
        visited = [[False for _ in range(col)] for _ in range(row)]

        for i in range(row):
            for j in range(col):
                if grid[i][j] == '1' and visited[i][j] == False:
                    self.bfs(i,j,visited,grid)
                    count += 1
            
        return count